proxy/upgrade.go

```diff
@@ -0,0 +1,195 @@
+package proxy
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// isUpgradeRequest returns true if and only if there is a "Connection"
+// key with the value "Upgrade" in Headers of the given request.
+func isUpgradeRequest(req *http.Request) bool {
```

Comment:
I would like to use it with SPDY/3.1. 
I am ok to have an --experimental-upgrade flag

1)
If you want to apply the filters on each request and response, you have to dissect all upgradeable protocols. That means writing protocol dissectors for each protocol. I think it should not be done within skipper.

2)
The problem with hijacking is that we do not own the sockets anymore. 

3)
Maybe there is also no http in between anymore so filters does not make sense to apply.